pkg/providers/cloudstack/cloudstack.go

```diff
@@ -424,28 +420,19 @@ func (p *cloudstackProvider) SetupAndValidateCreateCluster(ctx context.Context,
 }
 
 func (p *cloudstackProvider) SetupAndValidateUpgradeCluster(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec) error {
-	err := p.validateEnv(ctx)
-	if err != nil {
-		return fmt.Errorf("failed setup and validations: %v", err)
+	if err := p.validateEnv(ctx); err != nil {
+		return fmt.Errorf("validating environment variables: %v", err)
 	}
 
-	cloudStackClusterSpec := NewSpec(clusterSpec, p.machineConfigs, p.datacenterConfig)
-	if err := p.validator.validateCloudStackAccess(ctx); err != nil {
-		return err
-	}
-	if err := p.validator.ValidateCloudStackDatacenterConfig(ctx, p.datacenterConfig); err != nil {
-		return err
-	}
-	if err := p.validator.ValidateClusterMachineConfigs(ctx, cloudStackClusterSpec); err != nil {
-		return err
+	if err := p.validateClusterSpec(ctx, clusterSpec); err != nil {
+		return fmt.Errorf("validating cluster spec: %v", err)
 	}
 
 	if err := p.setupSSHAuthKeysForUpgrade(); err != nil {
-		return fmt.Errorf("failed setup and validations: %v", err)
+		return fmt.Errorf("setting up SSH keys: %v", err)
 	}
 
-	err = p.validateMachineConfigsNameUniqueness(ctx, cluster, clusterSpec)
-	if err != nil {
+	if err := p.validateMachineConfigsNameUniqueness(ctx, cluster, clusterSpec); err != nil {
```

Comment:
This is just a code clean up change. 
from 
```
err = func1()
if err != nil {}
```
to 
```
if err := func1(); err != nil {}
```
If you think validateMachineConfigsNameUniqueness is a part of validateClusterSpec then I should be handled in a separate PR because this pattern shows up in vsphere provider as well.

Reply:
I don't think vsphere provider code is a good reference. But yea separate pr sounds good to me.